[docs] Script for installing Deckhouse in kind

z9r5 opened the pull request:
## Description
Script for installing Deckhouse in kind on macOs or Debian based distros.

## Changelog entries
```changes
section: docs
type: feature
summary: Script for installing Deckhouse in kind on macOs or Debian based distros.
impact_level: low
```

deckhouse-BOaTswain commented:
Aye, aye, @z9r5. I've started the workflow for label 'deploy/web/stage'...
<!-- workflow_start -->

deckhouse-BOaTswain commented:
Aye, aye, @z9r5. I've started the workflow for label 'deploy/web/stage'...
<!-- workflow_start -->

deckhouse-BOaTswain commented:
Aye, aye, @z9r5. I've started the workflow for label 'deploy/web/stage'...
<!-- workflow_start -->

deckhouse-BOaTswain commented:
Aye, aye, @z9r5. I've started the workflow for label 'deploy/web/stage'...
<!-- workflow_start -->

deckhouse-BOaTswain commented:
Aye, aye, @z9r5. I've started the workflow for label 'deploy/web/stage'...
<!-- workflow_start -->

deckhouse-BOaTswain commented:
Aye, aye, @z9r5. I've started the workflow for label 'deploy/web/stage'...
<!-- workflow_start -->

deckhouse-BOaTswain commented:
Aye, aye, @z9r5. I've started the workflow for label 'deploy/web/stage'...
<!-- workflow_start -->

diafour commented on tools/kind-d8.sh:
```diff
@@ -0,0 +1,515 @@
+
+  echo "Creating Deckhouse installation config file (${CONFIG_DIR}/config.yml)..."
+  cat <<EOF >${CONFIG_DIR}/config.yml
+# Go to https://deckhouse.io/en/gs/kind/step4.html for comments to the file.
```
```suggestion
# InitConfiguration  documentation: https://deckhouse.io/en/gs/kind/step4.html
```

diafour commented on tools/kind-d8.sh:
```diff
@@ -0,0 +1,515 @@
+  D8_EE_ACCESS_STRING=$(echo -n "{\"auths\": { \"$D8_REGISTRY_ADDRESS\": { \"username\": \"license-token\", \"password\": \"$1\", \"auth\": \"$auth_part\"}}}" | base64 -w0)
+
+  if [ "$?" -ne "0" ]; then
+    echo "Error generation container registry access string ofr Deckhouse Enterprise Edition"
```
```suggestion
    echo "Error generation container registry access string for Deckhouse Enterprise Edition"
```

diafour commented on tools/kind-d8.sh:
```diff
@@ -0,0 +1,515 @@
+"
+  else
+    printf "
+You can now open URL http://grafana-127-0-0-1.nip.io/ in your browser and provide the following credentials:
```
```suggestion
Provide following credentials to access Grafana at http://grafana-127-0-0-1.nip.io/ :
```

z9r5 replied:
Fixed

diafour commented on tools/kind-d8.sh:
```diff
@@ -0,0 +1,515 @@
+
+  install_show_credentials >>${CONFIG_DIR}/info.txt
+
+  echo "You can find the information provided above in ${CONFIG_DIR}/info.txt file." >>${CONFIG_DIR}/info.txt
+
+  cat ${CONFIG_DIR}/info.txt
```
```suggestion
  cat ${CONFIG_DIR}/info.txt
  echo
  echo "The information above is saved to ${CONFIG_DIR}/info.txt file."
```

diafour reviewed (dismissed):
:tada: great work!

deckhouse-BOaTswain commented:
Aye, aye, @z9r5. I've started the workflow for label 'deploy/web/stage'...
<!-- workflow_start -->

nabokihms commented on tools/kind-d8.sh:
```diff
@@ -0,0 +1,502 @@
+Don't forget that the default kubectl context has been changed to 'kind-${KIND_CLUSTER_NAME}'.
+
+Run %s --context %s cluster-info to see cluster info.
+Run %s delete cluster --name %s to remove cluster.
```
```suggestion
Run '%s delete cluster --name %s' to remove cluster.
```

nabokihms commented on tools/kind-d8.sh:
```diff
@@ -0,0 +1,502 @@
+
+Don't forget that the default kubectl context has been changed to 'kind-${KIND_CLUSTER_NAME}'.
+
+Run %s --context %s cluster-info to see cluster info.
```
```suggestion
Run '%s --context %s cluster-info' to see cluster info.
```

nabokihms commented on tools/kind-d8.sh:
```diff
@@ -0,0 +1,502 @@
+
+    if [ "$retries_count" -ge "$retries_max" ]; then
+      echo
+      echo "Timeout waiting for the creation of Ingress controller!"
```
It would be best to output the last error here, or maybe the whole output of the kubectl -n d8-ingress-nginx get ds -o wide command output.

z9r5 replied:
Fixed.
Added more information. Here is an example of the output.

```
Waiting for the Ingress controller to be ready....
Timeout waiting for the creation of Ingress controller!

Here is the output of the 'kubectl -n d8-ingress-nginx get ds controller-nginx -o wide' command:
NAME               DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR   AGE   CONTAINERS                                     IMAGES                                                                                                                                                                                                                                                                                                                          SELECTOR
controller-nginx   1         1         1       1            1           <none>          34m   controller,protobuf-exporter,kube-rbac-proxy   registry.deckhouse.io/deckhouse/ce:191932fcff3a7cab75b79c08913f175456d5908c326df73c567f0b16-1646204184061,registry.deckhouse.io/deckhouse/ce:6de8c6e5b8ac80a87f2574a4f0ebcb62df4ce2542434a39318a013f6-1633631210461,registry.deckhouse.io/deckhouse/ce:a4506c2aa962611cf1858c774129d2a4f233502ecc376929aa97b9f5-1639403210069   app=controller,name=nginx

Here is the output of the 'kubectl -n d8-ingress-nginx get po -l app=controller' command:
NAME                     READY   STATUS    RESTARTS   AGE
controller-nginx-k2l6d   3/3     Running   0          34m

If the controller-nginx Pod is in the ContainerCreating status, you most likely have a slow connection. If so, wait a little longer until the controller-nginx Pod becomes Ready. After that, run the following command to get the admin password for Grafana: 'kubectl --context kind-d8 -n d8-system get cm deckhouse -o jsonpath="{.data.prometheus}" | grep password' (in case of success Grafana will be available at http://grafana-127-0-0-1.nip.io/).
```

nabokihms commented on tools/kind-d8.sh:
```diff
@@ -0,0 +1,502 @@
+  fi
+}
+
+ingress_check() {
```
You can made it a `bootstrap-script` to remove kubectl dependency. Could you please discuss it with @name212 ?

nabokihms replied:
Never mind, this is not possible. I thought that we could use the master node kubectl, but it looks like there is no option (or options are too complicated).

nabokihms commented on tools/kind-d8.sh:
```diff
@@ -0,0 +1,502 @@
+    prometheusEnabled: true
+    prometheus:
+      longtermRetentionDays: 0
+    monitoringKubernetesControlPlaneEnabled: true
```
I think we have a problem with kind installation. The Grafana is pretty empty right after the installation. I did not manage to find a working dashboard in the first five clicks. Maybe we need to enable a monitoring-kubernetes module also.

z9r5 replied:
Ok. I've enabled the `monitoring-kubernetes` module. I see that there are dashboards with data.

z9r5 commented:
@nabokihms There is a problem with the ebpf-exporter. An investigation is needed.
```
# kubectl -n d8-monitoring logs ebpf-exporter-wvrxn -c ebpf-exporter
sh: 1: modprobe: not found
chdir(/lib/modules/5.13.0-41-generic/build): No such file or directory
2022/05/23 04:10:15 Error attaching exporter: error compiling module for program "oomkill"
```

nabokihms commented:
@z9r5 this should be fixed in the main branch.

nabokihms commented:
The last time I tested this script, it timed out. I think this is because I do not have enough resources on my PC. 16Gb ram is not enough to run IDE and Deckhouse. Let's investigate this problem or write how many resources a user needs to install Deckhouselocally.

nabokihms reviewed (changes requested):
Btw, DCO check did not pass.

deckhouse-BOaTswain commented:
Aye, aye, @z9r5. I've started the workflow for label 'deploy/web/stage'...
<!-- workflow_start -->

deckhouse-BOaTswain commented:
Aye, aye, @z9r5. I've started the workflow for label 'deploy/web/stage'...
<!-- workflow_start -->

deckhouse-BOaTswain commented:
Aye, aye, @z9r5. I've started the workflow for label 'deploy/web/stage'...
<!-- workflow_start -->

z9r5 commented:
> The last time I tested this script, it timed out. I think this is because I do not have enough resources on my PC. 16Gb ram is not enough to run IDE and Deckhouse. Let's investigate this problem or write how many resources a user needs to install Deckhouselocally.

I've tested the script, and it works if there is at least 4 Gb of RAM. I've added a memory check:
 - if the script can get the total memory value and the value is less than 4GB, it prints the error and exits.
 - if the script can't get the total memory value, it prints a warning, waits to press a key, and continues.
- I didn't test the changes on macOS - can you check the script gets total memory in the right way? It should print no warning if you have more than 4 GB RAM.

I've added requirements for PC on the GS page.

Also, I've added the `--channel` parameter to specify the release channel (for debugging purposes mostly).

P.S. Deployed to stage.

z9r5 commented:
@nabokihms

deckhouse-BOaTswain commented:
Aye, aye, @z9r5. I've started the workflow for label 'deploy/web/stage'...
<!-- workflow_start -->

diafour commented on tools/kind-d8.sh:
```diff
@@ -0,0 +1,568 @@
+          D8_RELEASE_CHANNEL_NAME="$2"
+          D8_RELEASE_CHANNEL_TAG=$(echo ${D8_RELEASE_CHANNEL_NAME} | sed 's/EarlyAccess/early-access/; s/RockSolid/rock-solid/' | tr '[:upper:]' '[:lower:]')
+        else
+          echo "Incorrect release channel."
```
```suggestion
          echo "Incorrect release channel. Use Alpha, Beta, EarlyAccess, Stable or RockSolid"
```

diafour commented on tools/kind-d8.sh:
```diff
@@ -0,0 +1,568 @@
+KIND_INSTALL_DIRECTORY=$CONFIG_DIR
+KIND_PATH=kind
+KIND_CLUSTER_NAME=d8
+KIND_VERSION=v0.13.0
```
```suggestion
KIND_VERSION=v0.14.0
```

diafour commented on tools/kind-d8.sh:
```diff
@@ -0,0 +1,568 @@
+# limitations under the License.
+
+CONFIG_DIR=~/.kind-d8
+KIND_IMAGE=kindest/node:v1.21.12@sha256:6f19412d7d7c4156b3302d4de16168227173f155271be69250893e5c8585cb86
```
```suggestion
KIND_IMAGE=kindest/node:v1.21.12@sha256:f316b33dd88f8196379f38feb80545ef3ed44d9197dca1bfd48bcb1583210207
```

deckhouse-BOaTswain commented:
Aye, aye, @z9r5. I've started the workflow for label 'deploy/web/test'...
<!-- workflow_start -->

nabokihms commented:
Let's merge this whenever you are ready.
